source/common/unified_memory_allocation/include/uma/memory_provider.h

```diff
@@ -67,23 +67,25 @@ enum uma_result_t umaMemoryProviderFree(uma_memory_provider_handle_t hProvider,
 ///        independent way to return a provider specific result.
 ///
 /// \details
-///     - The string returned via the ppMessage is a NULL terminated C style
-///       string.
-///     - The string returned via the ppMessage is thread local.
-///     - The memory in the string returned via the ppMessage is owned by the
-///       adapter.
-///     - The application may call this function from simultaneous threads.
-///     - The implementation of this function should be lock-free.
+/// * Implementations *must* store the message and error code in thread-local
+///   storage prior to returning UMA_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC.
 ///
+/// * The message and error code storage is will only be valid if a previously
+///   called entry-point returned UMA_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC.
+///
+/// * The memory pointed to by the C string returned in `ppMessage` is owned by
+///   the adapter and *must* be null terminated.
+///
+/// * The application *may* call this function from simultaneous threads.
+///
+/// * The implementation of this function *should* be lock-free.
 /// \param hProvider handle to the memory provider
 /// \param ppMessage [out] pointer to a string containing provider specific
 ///        result in string representation
-/// \return UMA_RESULT_SUCCESS if the result being reported is to be considered
-///         a warning. Any other result code returned indicates that the
-///         adapter specific result is an error.
-enum uma_result_t
-umaMemoryProviderGetLastResult(uma_memory_provider_handle_t hProvider,
-                               const char **ppMessage);
+/// \param pError [out] pointer to an integer where the adapter specific error code will be stored
+void umaMemoryProviderGetLastNativeError(uma_memory_provider_handle_t hProvider,
```

Comment:
what if some pool uses several memory providers, e.g. one for data and another one for metadata?
what will be the user flow?

Reply:
That's a good point. User might need to know which (metadata or data) provider failed. I think we need to extend error codes to specify DATA_PROVIDER_ERROR and METADATA_PROVIDER_ERROR.

Reply:
I do not like the idea of introducing provider type-specific errors, especially if we consider changing the way how we specify provider type in the pool.

We can think of storing the handle to the provider that failed and remove `hProvider` argument from the `umaMemoryProviderGetLastNativeError` function.

In general, I would like to see some code-snippet that demonstrates how these API suppose to be used by client (e.g. UR).

Reply:
I've extended the error codes in https://github.com/oneapi-src/unified-runtime/pull/641

Reply:
If we assume that `umaMemoryProviderGetLastNativeError` can only be called by the memory pool user, then we could probably then we could probably replace the `hProvider` handle with a `pool` handle. I don't think we can remove the handle entirely because we need to know which function to call (from memory_provider_ops that is associated with that handle). 

Still I think `umaMemoryProviderGetLastNativeError` might also be useful **inside** a memory pool implementation (e.g. when a pool always uses some specific metadata provider).

The code would look similar to what I proposed in the comment above (just with extra switch for metadata/data provider):

void* ptr = umaPoolMalloc(hPool, size);
if (!ptr) {
    uma_result_t ret = umaPoolGetLastAllocationError(hPool);
    uma_memory_provider_handle_t provider;
    if (ret == UMA_RESULT_ERROR_DATA_MEMORY_PROVIDER_SPECIFIC) {
        provider = umaPoolGetDataMemoryProvider(hPool):
    } else if (ret == UMA_RESULT_ERROR_METADATA_MEMORY_PROVIDER_SPECIFIC {
        provider = umaPoolGetMetadataMemoryProvider(hPool):
    } else {
        ....
    }
    const char *msg;
    int32_t native_error;
    umaMemoryProviderGetLastNativeError(provider, &msg, &native_error);
    // cast native_error to ze_result_t and handle it
}

Reply:
> Still I think umaMemoryProviderGetLastNativeError might also be useful inside a memory pool implementation (e.g. when a pool always uses some specific metadata provider)

I agree with that.

But I would suggest to simplify the code snippet you provided above to the following:

```
void* ptr = umaPoolMalloc(hPool, size);
if (!ptr) {
    uma_result_t ret = umaPoolGetLastAllocationError(hPool);
    uma_memory_provider_handle_t provider;
    if (ret == UMA_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC) {
        const char *msg;
        int32_t native_error;
        // Why the pool internally cannot remember the failed memory provider?
        umaMemoryProviderGetLastNativeError(hPool, &msg, &native_error);
        // cast native_error to ze_result_t and handle it
    }
}
```

Why the pool internally cannot remember the failed memory provider? Such an approach does not require to introduce `UMA_RESULT_ERROR_DATA_MEMORY_PROVIDER_SPECIFIC` and `UMA_RESULT_ERROR_METADATA_MEMORY_PROVIDER_SPECIFIC` error codes.

Also, another question came to my mind. What about external heap manager implementations (like jemalloc), does such approach requires any changes in the heap mgr implementation itself?

Reply:
> Why the pool internally cannot remember the failed memory provider?

Well, it can remember it, but where would you pass that provider handle if `umaMemoryProviderGetLastNativeError` no longer accepts it? I wouldn't want to introduce some provider interface specifically for use inside the memory pool implementation. 

> Also, another question came to my mind. What about external heap manager implementations (like jemalloc), does such approach requires any changes in the heap mgr implementation itself?

For jemalloc we would just record the appropriate error in callbacks (.e.g `hooks.alloc`) - that's the place where we would use a memory provider. I don't think jemalloc would use metadata memory provider but that's fine.

Reply:
Also, I think it is useful to know which provider failed: whether it's metadata or data provider. If you just call `umaMemoryProviderGetLastNativeError(hPool, &msg, &native_error);` as you suggest you don't know which provider failed and you don't know how to cast the native_error (since those 2 providers can have different native error codes).

Reply:
I got your point. But the only thing I want to fix is to get rid of these two error codes (`UMA_RESULT_ERROR_DATA_MEMORY_PROVIDER_SPECIFIC` and `UMA_RESULT_ERROR_METADATA_MEMORY_PROVIDER_SPECIFIC`) and have only one `UMA_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC`.

Reply:
The reason I want to avoid it is related to [this comment](https://github.com/oneapi-src/unified-runtime/pull/641/files#r1244132297). The proposed pool creation API is very specific to current USM needs.
Imagine we will change the pool creation API in the future then ERROR handling code by the UMF clients has to be changed as well.

Reply:
But then how would you know which provider failed? (i.e. which provider to pass to the `umaMemoryProviderGetLastNativeError`)?

Reply:
If we decide not to merge https://github.com/oneapi-src/unified-runtime/pull/641 we need some other way of specyfing which memory provider is the one that failed. The simplest approach would be just to have only one provider per pool (as we had originally).

Reply:
Anyway, this PR does not introduce UMA_RESULT_ERROR_DATA_MEMORY_PROVIDER_SPECIFIC nor UMA_RESULT_ERROR_METADATA_MEMORY_PROVIDER_SPECIFIC so I think this discussion can continue on this thread: https://github.com/oneapi-src/unified-runtime/pull/641/files#r1244132297

Can this PR be merged or do you have any other comments?

Reply:
In general, to interpret the native error code you should know the concrete implementation referred by the memory provider handle.

> Also, I think it is useful to know which provider failed: whether it's metadata or data provider.

And it does not matter whether it DATA or METADATA memory provider. You just need to know which memory provider failed. For example, imagine the following situation:

1. You have two memory providers (different implementations - different ops structure) and pass them to `umaPoolCreate` function. 
2. `umaPoolMalloc` call failed because one of the memory providers failed. Does it really matter whether failed memory provider was used as DATA or METADATA? One obvious solution could be just to have `umaPoolGetLastFailedProvider`  function to get failed provider. It will help to avoid `if-else if-...-else` block.

But an even better idea is to look how similar problem solved in STL: [`std::error_code`](https://en.cppreference.com/w/cpp/error/error_code)
Also, it would be useful to understand how clients are going to use this native error? Do we have a code snippet from UR perspective?

Reply:
Done. I've implemented `umaGetLastFailedProvider` as a standalone function that does not accept a `pool` handle so that it can be implemented in memory_provider.c (as opposed to in each memory pool separately).